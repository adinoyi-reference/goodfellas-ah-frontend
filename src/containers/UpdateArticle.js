import React, { Component } from 'react';
import { connect } from 'react-redux';
import swal from 'sweetalert';
import Editor from 'react-medium-editor';
import 'medium-editor/dist/css/medium-editor.css';
import 'medium-editor/dist/css/themes/default.css';
import ImageUploader from '../components/articles/imageUpload';
import updateArticle from '../actions/updateArticle';
import { getAnArticle } from '../actions/articleActions';
import '../styles/views/createArticles.scss';
import { Loader } from '../components/shared/Loading';
import icons from '../assets/icons.svg';


export class UpdateArticles extends Component {
  state = {
    title: '',
    body: '',
    imageLoad: false,
  }

  componentDidMount = () => {
    const { getAnArticle: fetchArticle, match } = this.props;
    fetchArticle(match.params.slug);
  }

  componentDidUpdate = (prevProps) => {
    const {
      status, article, history, match,
    } = this.props;
    if (prevProps.status.success !== status.success && status.success) {
      swal('Good job!', 'Article Updated Successfully!', 'success');
      setTimeout(() => history.push(`/articles/${match.params.slug}`), 3000);
    } else if (prevProps.status.error !== status.error && status.error) {
      swal('Error!', 'Something Went Wrong!', 'error');
    }
    if (prevProps.article !== article) this.setState({ title: article.title, body: article.body });
  }

  handleChange = (event) => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  }

  generateImageTag = (url) => `<img class="inline-image" src="${url}" />`


  imageUploaded = (url) => {
    const imageHtml = this.generateImageTag(url);
    this.setState((prevState) => ({ body: prevState.body + imageHtml, imageLoad: false }));
  }

  handleSubmit = async (event) => {
    event.preventDefault();
    const { title, body } = this.state;
    if (!title || title === '<p><br></p>') return swal('Please add a title', 'Your article needs to have a title', 'warning');
    if (!body || body === '<p><br></p>') { return swal('Please add a body', 'The body of your article cannot be empty!', 'warning'); }
    const { match } = this.props;
    const articlePayload = {
      ...this.state,
      description: body.split(' ').slice(0, 10).join(' '),
    };
    // eslint-disable-next-line react/destructuring-assignment
    this.props.updateArticle(articlePayload, match.params.slug);
  }

  render() {
    const handleEditorChange = (text, key) => this.setState({ [key]: text });
    const { imageUploadStatus, articleLoading, articleError } = this.props;
    const { title, body } = this.state;
    if (articleLoading) return <Loader />;
    if (articleError) {
      return (
        <div className="no-record centralizer">
          <svg className="icon">
            <use xlinkHref={`${icons}#sad`} />
          </svg>&nbsp;&nbsp;
          <span>{articleError}</span>
        </div>
      );
    }
    return (

      <div className="container article-body">
        <div className="articles-card">
          <div className="article-buttons">
            <button className="btn article-whitebutton" type="submit" onClick={this.handleSubmit}>
            Update
            </button>
          </div>
          <ImageUploader imageUploaded={this.imageUploaded} />

          <form>
            <textarea id="title" value={title} onChange={(e) => handleEditorChange(e.target.value, 'title')} />
            {imageUploadStatus.loading ? <Loader /> : (
              <Editor
                name="body"
                className="editable"
                text={body}
                onChange={(e) => handleEditorChange(e, 'body')}
                options={{
                  placeholder: false,
                  toolbar: {
                    buttons: ['bold', 'italic', 'underline', 'strikethrough', 'quote', 'anchor', 'h2', 'h3', 'orderedlist'],
                  },
                }}
              />
            )}
          </form>

        </div>
        <div className="article-fixed-buttons">
          <button className="btn article-fixed-whitebutton" type="submit" onClick={this.handleSubmit}>
            Update
          </button>
        </div>
      </div>
    );
  }
}


const mapStateToProps = ({
  imageUploadReducer: { status: imageUploadStatus },
  updateArticleReducer: { status, updatedArticle },
  articles: { article, articleLoading, articleError },
}) => ({
  status,
  updatedArticle,
  imageUploadStatus,
  article,
  articleLoading,
  articleError,
});

export default connect(mapStateToProps, { updateArticle, getAnArticle })(UpdateArticles);
