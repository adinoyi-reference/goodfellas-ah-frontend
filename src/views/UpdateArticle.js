import React, { Component } from 'react';
import { connect } from 'react-redux';
import swal from 'sweetalert';
import Editor from 'react-medium-editor';
import 'medium-editor/dist/css/medium-editor.css';
import 'medium-editor/dist/css/themes/default.css';
import ImageUploader from '../components/articles/imageUpload';
import updateArticle from '../actions/updateArticle';
import '../styles/views/createArticles.scss';
import { Loader } from '../components/shared/Loading';
import Body from '../components/layout/DefaultLayout';

export class UpdateArticles extends Component {
  state = {
    title: '',
    body: '',
    imageLoad: false,
  }

  componentDidUpdate(prevProps) {
    const { status } = this.props;
    if (prevProps.status.success !== status.success && status.success) {
      swal('Good job!', 'Article Updated Successfully!', 'success');
    } else if (prevProps.status.error !== status.error && status.error) {
      swal('Error!', 'Something Went Wrong!', 'error');
    }
  }

  handleChange = (event) => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  }

  generateImageTag = (url) => `<img class="inline-image" src="${url}" />`


  imageUploaded = (url) => {
    const imageHtml = this.generateImageTag(url);
    this.setState((prevState) => ({ body: prevState.body + imageHtml, imageLoad: false }));
  }

  handleSubmit = async (event) => {
    event.preventDefault();
    const { body } = this.state;
    const articlePayload = {
      ...this.state,
      description: body.split(' ').slice(0, 3).join(' '),
    };
    // eslint-disable-next-line react/destructuring-assignment
    this.props.updateArticle(articlePayload);
  }

  render() {
    const handleEditorChange = (text, key) => this.setState({ [key]: text });
    const { imageUploadStatus } = this.props;
    const { title, body } = this.state;
    return (
      <Body>
        <div className="article-body">
          <div className="articles-card">
            <div className="article-buttons">
              <button className="btn article-whitebutton" type="submit" onClick={this.handleSubmit}>
            Update
              </button>
            </div>
            <ImageUploader imageUploaded={this.imageUploaded} />

            <form>
              <Editor
                name="title"
                id="title"
                data-placeholder="Title"
                text={title}
                onChange={(e) => handleEditorChange(e, 'title')}
                options={{
                  toolbar: {
                    static: true,
                    sticky: false,
                    buttons: ['bold', 'italic', 'underline', 'strikethrough', 'quote', 'anchor', 'h2', 'h3', 'orderedlist'],
                    updateOnEmptySelection: true,
                  },
                }}
              />
              {imageUploadStatus.loading ? <Loader /> : (
                <Editor
                  name="body"
                  className="editable"
                  data-placeholder="What would you like to talk about?"
                  text={body}
                  onChange={(e) => handleEditorChange(e, 'body')}
                  options={{
                    toolbar: {
                      activeButtonClass: 'medium-editor-button-active',
                      allowMultiParagraphSelection: true,
                      static: true,
                      sticky: false,
                      buttons: ['bold', 'italic', 'underline', 'strikethrough', 'quote', 'anchor', 'h2', 'h3', 'orderedlist'],
                      updateOnEmptySelection: true,
                    },
                  }}
                />
              )}
            </form>

          </div>
          <div className="article-fixed-buttons">
            <button className="btn article-fixed-whitebutton" type="submit" onClick={this.handleSubmit}>
            Update
            </button>
          </div>
        </div>
      </Body>
    );
  }
}


const mapStateToProps = ({
  imageUploadReducer: { status: imageUploadStatus },
  updateArticleReducer: { status, updatedArticle },
}) => ({
  status,
  updatedArticle,
  imageUploadStatus,
});

export default connect(mapStateToProps, { updateArticle })(UpdateArticles);
