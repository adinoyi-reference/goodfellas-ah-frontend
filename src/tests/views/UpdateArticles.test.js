import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import UpdateArticle from '../../views/UpdateArticle';

let wrapped;

beforeEach(() => {
  wrapped = shallow(<UpdateArticle />);
});


describe('Update Article UI', () => {
  describe('render features', () => {
    test('view should render as expected', () => {
      const tree = toJson(wrapped);
      expect(tree).toMatchSnapshot();
    });
  });
});
