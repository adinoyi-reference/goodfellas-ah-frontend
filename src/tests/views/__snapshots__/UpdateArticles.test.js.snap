// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update Article UI render features view should render as expected 1`] = `
<div>
  <_default>
    <Connect(UpdateArticles) />
  </_default>
</div>
`;
